### DISCUSSION TITLE
Why is ` __attribute__((used))` used with the function `void initialize_tcb(struct tcpcb* tp)`

### DISCUSSION THREAD

**Discussion by romacdon** [openthread/openthread#10715] [Q&A]
Anyone know why` __attribute__((used))` is used with the function `void initialize_tcb(struct tcpcb* tp)` in `tcp_subr.c`?  It causes an issue when used with link time optimizations and TCP is not enable.

`openthread/third_party/tcplp/bsdtcp/tcp_subr.c:96: undefined reference to tcplp_sys_get_ticks'`

@samkumar perhaps you might know the answer.

**Comment by romacdon** [comment]
@jwhui / @abtink - any objections if I open a PR to remove this attribute?

**Reply by abtink** [reply]
This seems to be in third-party source code for `tcplp`. Not sure why `__attribute__((used))` is used. We can update this in the OT repo, but it may be better to check if the upstream version can be updated first.

@samkumar, it would be great to hear your thoughts.

**Reply by samkumar** [reply]
No objections on my end (see my more detailed comment below).

**Comment by romacdon** [comment]
Thanks for replying @abtink .  I was thinking the same (about needing to update the upstream version) but then I noticed other PRs that were directly changing the tcplp code in the openthread repo.  For example. https://github.com/openthread/openthread/pull/8578.

**Reply by abtink** [reply]
I see. Interesting. Not the best model, but if there is precedence I guess we can change other things as well.
I did check the code and I cannot tell why `__attribute__((used))` is added for `initialize_tcb()`?

**Comment by romacdon** [comment]
Thanks @abtink.  I'll get a PR in place for this.  Also, just a heads up I also want to fix the compilation warnings that occur when you remove these compiler options when compiling the tcplp code, ""-Wno-sign-compare" and "-Wno-unused-parameter".  I'll create a separate PR for those changes.

**Reply by samkumar** [reply]
I'm curious what the changes to remove the `-Wno-sign-compare` and `-Wno-unused-parameter` look like. I suspect that some of the code that causes those warnings is the FreeBSD code that TCPlp is based on. I was reluctant to make major changes to it when originally developing TCPlp, but am potentially open to it now (particularly if the changes required to make this work aren't too sweeping).

**Reply by romacdon** [reply]
I'll create a PR with these changes in the next day or two.

**Comment by samkumar** [comment]
Thanks @romacdon for pointing this out. I would have to look back at the history to find out how this annotation got there. I originally developed TCPlp as a research prototype to work with embedded operating systems based on TinyOS, and soon after with RIOT OS, and I suspect that I put in `__attribute__((used))` at some point to get the build working with one of those systems. I'll look back at the commit history to see what exactly happened.

I can confirm that I'm not aware of any good reason to have `__attribute__((used))` in the current version of TCPlp for OpenThread. It was present in an older version of TCPlp, and I didn't think to remove it when I was working on the OpenThread port. So, I would be supportive of a pull request to OpenThread to delete the `__attribute__((used))` annotation in TCPlp, particularly since it causes problems for the build.

**Reply by romacdon** [reply]
Thanks @samkumar. I'll get a PR in place today.

**Reply by romacdon** [reply]
Just opened the PR - https://github.com/openthread/openthread/pull/10726

**Comment by romacdon** [comment]
Although the question was not answered I am closing this discussion as the attribute has been removed in this PR - https://github.com/openthread/openthread/pull/10726
